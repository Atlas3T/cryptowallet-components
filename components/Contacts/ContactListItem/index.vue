<template>
  <div
    class="contact-list-item"
    @click.prevent="clickHandler"
  >
    <div class="columns is-mobile no-margin-bottom">
      <div class="column is-2">
        <div class="circle" />
      </div>
      <div class="column">
        <div class="columns is-mobile no-margin-bottom">
          <div class="column is-7 no-padding-bottom">
            {{ contact.displayName }}
          </div>
        </div>
        <div class="wallet-address">
          <div v-if="contact.address">
            {{ contact.address }}
          </div>
          <div v-else>
            {{ $t('noWalletForContact') }}
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'ContactListItem',
  props: {
    contact: {
      type: Object,
      required: true,
    },
    clickItemAction: {
      type: String,
      required: false,
    },
  },
  methods: {
    /**
     * Initiates sending the invitation
     */
    clickHandler() {
      switch (this.clickItemAction) {
        case 'app-invitation':
          this.invitation = new this.AppInvitation({ vm: this, contact: this.contact });
          this.invitation.send();
          break;
        default:
          return false;
      }
      return false;
    },
  },
};
</script>

<style scoped>
.contact-list-item {
  margin: 1rem 0;
}

.wallet-address {
  font-size: 0.7rem;
}

.no-padding-bottom {
  padding-bottom: 0;
}

.no-margin-bottom {
  margin-bottom: 0;
}

.circle {
  background: #ececec;
  border-radius: 100%;
  height: 2.5rem;
  width: 2.5rem;
  margin-top: 0.1rem;
}
</style>
