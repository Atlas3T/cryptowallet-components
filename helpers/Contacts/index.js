import Vue from 'vue';

/**
 * Handles the contacts import and saves them in the store
 */
export default function Contacts() {
  this.vm = new Vue();
  this.import = (contacts) => {
    this.contacts = contacts;
    this.vm.$once('contactsImported', () => { return this.success(); });
    this.vm.$once('contactsImportFailure', () => { return this.failure(); });
    this.vm.$once('contactsImportError', (event) => { return this.error(event.payload); });

    try {
      const options = new ContactFindOptions();
      options.multiple = true;
      const fields = ['*'];
      navigator.contacts.find(fields, (contactsArray) => {
        this.contacts.contacts = contactsArray;
        this.vm.$emit('contactsImported');
      }, () => {
        this.vm.$emit('contactsImportFailure');
      }, options);
    } catch (err) {
      this.vm.$emit('contactsImportError', { payload: err });
    }

    return this;
  };

  this.on = (event, callback) => {
    this[event] = callback;
    return this;
  };
}
