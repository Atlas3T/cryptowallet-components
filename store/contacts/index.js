export default {
  namespaced: true,
  state: {
    contacts: [
      {
        id: 1,
        displayName: 'Terry Tibbs',
        address: 'ncousohfodicwjdclwdnc',
      },
    ],
  },
};
