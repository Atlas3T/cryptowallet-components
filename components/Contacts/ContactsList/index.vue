<template>
  <section
    :class="{ isSearchingContacts: isSearchingContacts }"
    class="contacts-list"
  >
    <h1 class="section-h1">
      {{ $t('contacts') }}
    </h1>
    <div class="flex-scroll-area-wrapper">
      <q-scroll-area class="contacts-list-scroll-area">
        <div class="scroll-content">
          <ContactListItem
            v-for="contact in filteredContacts"
            :key="contact.displayName"
            :contact="contact"
            :click-item-action="clickItemAction"
          />
        </div>
      </q-scroll-area>
    </div>
  </section>
</template>

<script>
import { mapState } from 'vuex';
import ContactListItem from '@/components/Contacts/ContactListItem';

export default {
  name: 'ContactsList',
  components: {
    ContactListItem,
  },
  props: {
    clickItemAction: {
      type: String,
      required: false,
    },
  },
  computed: {
    ...mapState({
      isSearchingContacts: (state) => {
        return state.search.isSearchingContacts;
      },
      searchingContactsQueryString: (state) => {
        return state.search.searchingContactsQueryString;
      },
      contacts: (state) => {
        return state.contacts.contacts;
      },
    }),

    /**
     * Filters contact based on the search input value
     * @return {Array}
     */
    filteredContacts() {
      return this.contacts.filter((contact) => { return new RegExp(`^${this.searchingContactsQueryString}`, 'i').test(contact.displayName); });
    },
  },
};
</script>

<style scoped>
.section-h1 {
  color: white;
  font-size: 0.9rem;
  font-family: 'Montserrat-Light';
  background: #162f46;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #0a2338;
  letter-spacing: 0.05rem;
  text-transform: uppercase;
}

.flex-scroll-area-wrapper {
  flex: auto;
  flex-grow: 1;
  position: relative;
  overflow: hidden;
  height: 100%;
}

.contacts-list-scroll-area {
  flex-grow: 1;
  width: 100%;
  height: 100%;
  position: absolute;
}

.contacts-list {
  border-top: 2px solid #09233a;
  position: relative;
  background: #16324a;
  display: flex;
  flex-direction: column;
  transition: height 100ms ease-in-out;
  height: 100%;
}

.contacts-list.isSearchingContacts {
  height: 85vh;
}

.scroll-content {
  overflow: hidden;
  padding: 0 1rem;
}
</style>
