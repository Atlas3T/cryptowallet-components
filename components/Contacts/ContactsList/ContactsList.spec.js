import { shallowMount, mount } from '@vue/test-utils';
import ContactsList from '@/components/Contacts/ContactsList';
import ContactListItem from '@/components/Contacts/ContactListItem';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';

describe('ContactsList.vue', () => {
  let storeMocks;
  let wrapper;
  let store;

  function wrapperInit (options) {
    return shallowMount(ContactsList, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    wrapper = wrapperInit({ i18n, localVue, store: storeMocks.store });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a section with .contacts-list class', () => {
    expect(wrapper.contains('section.contacts-list')).toBe(true);
  });

  it('renders a h1 with .section-h1 class that contains "Contacts" string', () => {
    expect(wrapper.find('h1.section-h1').html().includes('Contacts')).toBe(true);
  });

  it('loops through the data and renders multiple ContactListItem components', () => {
    expect(wrapper.findAll(ContactListItem).length).toBeGreaterThan(0);
  });

  it('adds the .isSearchingContacts class if store has search.isSearchingContacts === true ', () => {
    expect(store.state.search.isSearchingContacts).toBe(false);
    expect(wrapper.contains('section.contacts-list.isSearchingContacts')).toBe(false);
    store.state.search.isSearchingContacts = true;
    expect(wrapper.contains('section.contacts-list.isSearchingContacts')).toBe(true);
    store.state.search.isSearchingContacts = false;
  });

  it('filters the contacts if store has search.searchingContactsQueryString', () => {
    wrapper = mount(ContactsList, { i18n, localVue, store: storeMocks.store });
    expect(store.state.contacts.contacts.length === wrapper.findAll(ContactListItem).length).toBe(true);
    store.state.search.searchingContactsQueryString = 'pseudorandomstring';
    expect(wrapper.findAll(ContactListItem).length).toBe(0);
    store.state.search.searchingContactsQueryString = 'Te';
    expect(wrapper.findAll(ContactListItem).length).toBe(1);
    expect(wrapper.findAll(ContactListItem).at(0).html().toLowerCase().includes('te')).toBe(true);
  });
});
