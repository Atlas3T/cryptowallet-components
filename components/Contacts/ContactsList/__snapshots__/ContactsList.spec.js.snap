// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsList.vue renders and matches snapshot 1`] = `
<section
  class="contacts-list"
>
  <h1
    class="section-h1"
  >
    Contacts
  </h1>
   
  <div
    class="flex-scroll-area-wrapper"
  >
    <qscrollarea-stub
      class="contacts-list-scroll-area"
      contentactivestyle="[object Object]"
      contentstyle="[object Object]"
      delay="1000"
      thumbstyle="[object Object]"
    >
      <div
        class="scroll-content"
      >
        <contactlistitem-stub
          contact="[object Object]"
        />
      </div>
    </qscrollarea-stub>
  </div>
</section>
`;
